client/verta/verta/runtime.py

```diff
@@ -0,0 +1,173 @@
+    For use within a model's ``predict()`` method to collect logging
+    context, which is stored in thread-local variables.
+
+    Parameters
+    ----------
+    key : str
+        String value to use as data label (key) in logging context.
+    value : Any
+        Any JSON serializable value you wish to include in the context logs.
+
+    Returns
+    -------
+    None
+
+    Examples
+    --------
+    .. code-block:: python
+       :emphasize-lines: 10,13
+
+        from verta import runtime
+
+        # Sample model code:
+        class MyModel(VertaModelBase):
+            def __init__(self, artifacts):
+                "ok"
+
+            def predict(self, data):
+                logs = {'some_stuff': 'I_care_about'}
+                runtime.log('my_logging_key', logs)
+                prediction_output = 'some_output'
+                more_logs = ['a', 'list', 'of', 'things']
+                runtime.log('things', more_logs)
+                return prediction_output
+
+    """
+    if _get_validate_flag():
+        validate_json(value)
+    local_context: Dict[str, Any] = _get_thread_logs()
+    local_context.update({key: value})
+    _set_thread_logs(local_context)
+
+
+class context:
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    Uses thread-local variables to store context in a thread-safe manner.
+
+    Parameters
+    ----------
+    validate : Optional[bool]
+        If true, each individual call to ``runtime.log('key', value)`` will
+        verify that the value provided is JSON serializable.
+
+    Attributes
+    ----------
+    logs : Dict[str, Any]
+        Dictionary of the current logging context.  If called after exiting
+        the context manager, the final complete log entry is returned.
+    """
+    def __init__(self, validate: Optional[bool] = False):
+        self.validate = validate
+        self.logs_dict = dict()
+
+    def __enter__(self):
+        """
+        Ensure an empty logging context to start and set validation flag.
+        """
+        _set_thread_logs(dict())
+        if self.validate:
+            _set_validate_flag(True)
+        return self
+
+    def __exit__(self, *args):
+        """
+        Capture the final complete log entry in a class variable, ensure an
+        empty logging context, and reset the validation flag upon exit.
+        """
+        self.logs_dict = _get_thread_logs()
+        _set_thread_logs(dict())
+        _set_validate_flag(False)
+
+    def logs(self) -> Dict[str, Any]:
+        """
+        Return the currently stored, thread-local logs from inside this
+        context manager. JSON serialization is attempted on the whole
+        dictionary and an error is raised on failure.
+
+        If called after exiting the context manager, the final complete
+        log entry is returned.
+        """
+        logs: Dict[str, Any] = self.logs_dict or _get_thread_logs()
+        validate_json(logs)
```

Comment:
Would you mind removing the validation step here? It does make sense on its own, but the Python Model Service will already propagate an error if the final logs can't be JSONified, and having this here as a redundancy can be costly for performance if the logs are large.

Reply:
That's a fair proposal, but after thinking about this some more I've realized that it might not be a good idea to have a check here at all.

Here's one way to look at this: If the user set `validate=True`, then we are checking each value for JSON serializability in `runtime.log()`, right? _Therefore_ the items here in `logs` would already be guaranteed to be JSON-serializable: We've already checked the values, and the keys already have to be simple strings. So, if `validate=True`, this check here is a performance hit with no additional value: Everything has already been validated.

And furthermore—commenting on performance—this check actually ends up being a marginally _worse_ equivalent to simply checking in `runtime.log()` (it has to process the full `logs` in memory, plus it has to check the keys). Which is another reason not to do it by default: It's a slower version of the thing we already decided against for being too slow.

Reply:
Changed my mind and just dropped it altogether.  If `validate` is set to `True`, every key and value has been validated already when added.  If `validate` is set to `False`, the PyMS will through the `TypeError` back down in the 400 response if bad values were given.

Reply:
Done